realm/realm-library/src/main/cpp/io_realm_internal_OsRealmConfig.cpp

```diff
@@ -132,16 +132,26 @@ JNIEXPORT void JNICALL Java_io_realm_internal_OsRealmConfig_nativeSetSchemaConfi
             static JavaMethod run_migration_callback_method(
                 env, get_shared_realm_class(env), "runMigrationCallback",
                 "(JLio/realm/internal/OsRealmConfig;Lio/realm/internal/SharedRealm$MigrationCallback;J)V", true);
-            JavaGlobalRef j_config_ref(env, j_config, true);
-            JavaGlobalRef j_migration_callback_ref(env, j_migration_callback, true);
-            config.migration_function = [j_config_ref, j_migration_callback_ref](SharedRealm old_realm,
+            // weak ref to avoid leaks caused by circular refs.
```

Comment:
What kind of circular refs? configs cannot reference each other?

Reply:
@cmelchior eg.:
you have a `CompactOnLaunch` as below:

```java
class MyCompactOnLaunch implements CompactOnLaunch {
    OsRealmConfig config;
}
```

Then this `compactOnLaunch` instance has been passed to the Object Store in the function block as a global ref.

```c++
struct Config {
//...
        ShouldCompactOnLaunchFunction should_compact_on_launch_function;
//...
}
```

The native `Config` instance won't be freed until the `OsRealmConfig` gets GCed. But since the `should_compact_on_launch_function` holding a global ref to the java `MyCompactOnLaunch` instance, it will never be GCed since java GC cannot detect if the global ref is still necessary in native code.

The real case could be more complicated than this, I think we should always use the weak ref in c++ lambda captures (unless there is a real need, but again, that makes things very complex.)

Reply:
Good point 👍